docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,765 @@
+GET /_cluster/state?filter_path=metadata.cluster_coordination.voting_config_exclusions
+--------------------------------------------------
+// CONSOLE
 
-The `cluster.name` allows to create separated clusters from one another.
-The default value for the cluster name is `elasticsearch`, though it is
-recommended to change this to reflect the logical group name of the
-cluster running.
+This list is limited in size by the following setting:
 
-include::discovery/azure.asciidoc[]
+`cluster.max_voting_config_exclusions`::
 
-include::discovery/ec2.asciidoc[]
+    Sets a limits on the number of voting configuration exclusions at any one
+    time.  Defaults to `10`.
+
+Since voting configuration exclusions are persistent and limited in number, they
+must be cleaned up. Normally an exclusion is added when performing some
+maintenance on the cluster, and the exclusions should be cleaned up when the
+maintenance is complete. Clusters should have no voting configuration exclusions
+in normal operation.
+
+If a node is excluded from the voting configuration because it is to be shut
+down permanently then its exclusion can be removed once it has shut down and
+been removed from the cluster. Exclusions can also be cleared if they were
+created in error or were only required temporarily:
+
+[source,js]
+--------------------------------------------------
+# Wait for all the nodes with voting configuration exclusions to be removed from
+# the cluster and then remove all the exclusions, allowing any node to return to
+# the voting configuration in the future.
+DELETE /_cluster/voting_config_exclusions
+# Immediately remove all the voting configuration exclusions, allowing any node
+# to return to the voting configuration in the future.
+DELETE /_cluster/voting_config_exclusions?wait_for_removal=false
+--------------------------------------------------
+// CONSOLE
+
+[float]
+[[cluster-state-publishing]]
+=== Cluster state publishing
+
+The master node is the only node in a cluster that can make changes to the
+cluster state. The master node processes one cluster state update at a time,
+applies the required changes and publishes the updated cluster state to all the
+other nodes in the cluster. Each node receives the publish message, acknowledges
+it, but does *not* yet apply it. If the master does not receive acknowledgement
+from enough master-eligible nodes within a certain time (controlled by the
+`cluster.publish.timeout` setting which defaults to 30 seconds) the cluster
+state change is rejected.
+
+Once enough nodes have responded, the cluster state is committed and a commit
+message is sent to all the nodes. The nodes then proceed to apply the new
+cluster state to their internal state. The master node waits for all nodes to
+respond, or until `cluster.publish.timeout` has elapsed, before starting to
+process the next update in the queue. The `cluster.publish.timeout` is measured
+from the moment the publishing started.
```

Comment:
It's a bit confusing. There are two messages "publish" and "commit". For "publish" message there is "cluster.publish.timeout". Publish is given 30 seconds to complete. Consider it takes X seconds. If X < 30 seconds, we're sending "commit" message. Does it mean "commit" has only (30-X) seconds to succeed?

Reply:
Conceptually both messages are involved in the publication process, see for instance `PublicationTransportHandler` or even `Publication`, although I will admit that one of the actions is named `internal:cluster/coordination/publish_state`. In Zen1 it was `internal:discovery/zen/publish/send`, which is perhaps a more helpful name for this question. Regardless, it's the whole publication process that's timing out after 30 seconds regardless of what state it's in.

So yes, each `internal:cluster/coordination/commit_state` only has whatever is left of the 30-second timeout after the `internal:cluster/coordination/publish_state` has completed.

Reply:
Can we re-word the description to make it clear that publication consists of two phases "publish" and "commit" and `cluster.publish.timeout` is the timeout for the whole publication?
Of course, it would be nice to rename "publish" to something else, because it's confusing, that first stage of "publication" is "publish". I'd expect it's the only stage because it has the same name. But we can leave as it is.

Reply:
Sure, I've reworded it to more carefully use "publication" to refer to the whole process: 1888c97fa6bd0b93a473cc5ebe8be6f6991ab6a4.